modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -480,6 +625,152 @@ public int getZoneId(String name) {
+                    }
+
+                    if (scaleUpRevision != null
+                            && (dataNodes.get(zoneId) == null || dataNodes.get(zoneId).scaleUpRevision() < scaleUpRevision)) {
+                        Map.Entry<Long, CompletableFuture<Void>> ceilingEntry =
+                                dataNodes.get(zoneId).revisionScaleUpFutures().ceilingEntry(scaleUpRevision);
+
+                        if (ceilingEntry != null) {
+                            topVerScaleUpFut = ceilingEntry.getValue();
+                        }
+
+                        if (topVerScaleUpFut == null) {
+                            topVerScaleUpFut = new CompletableFuture<>();
+
+                            dataNodes.get(zoneId).revisionScaleUpFutures().put(scaleUpRevision, topVerScaleUpFut);
+                        }
+                    } else {
+                        topVerScaleUpFut = completedFuture(null);
+                    }
+                } else {
+                    topVerScaleUpFut = completedFuture(null);
+                }
+
+                topVerScaleUpFut.thenAcceptAsync(ignored0 -> {
+                    CompletableFuture<Void> topVerScaleDownFut = null;
+
+                    synchronized (dataNodesMutex) {
+                        if (immediateScaleDown) {
+                            Map.Entry<Long, Long> scaleDownRevisionEntry = topVerAndScaleDownRevision.ceilingEntry(topVer);
+
+                            Long scaleDownRevision = null;
+
+                            if (scaleDownRevisionEntry != null) {
+                                scaleDownRevision = scaleDownRevisionEntry.getValue();
+                            }
+
+                            if (scaleDownRevision != null
+                                    && (dataNodes.get(zoneId) == null || dataNodes.get(zoneId).scaleDownRevision() < scaleDownRevision)) {
+                                Map.Entry<Long, CompletableFuture<Void>> ceilingEntry =
+                                        dataNodes.get(zoneId).revisionScaleDownFutures().ceilingEntry(scaleDownRevision);
+
+                                if (ceilingEntry != null) {
+                                    topVerScaleDownFut = ceilingEntry.getValue();
+                                }
+
+                                if (topVerScaleDownFut == null) {
+                                    topVerScaleDownFut = new CompletableFuture<>();
+
+                                    dataNodes.get(zoneId).revisionScaleDownFutures.put(scaleDownRevision, topVerScaleDownFut);
+                                }
+                            } else {
+                                topVerScaleDownFut = completedFuture(null);
+                            }
+                        } else {
+                            topVerScaleDownFut = completedFuture(null);
+                        }
+                    }
+
+                    topVerScaleDownFut.thenAcceptAsync(ignored1 -> {
```

Comment:
Why do we need this code? Why we do this only got scale down?

Reply:
I rework it. Seems now it is more clear.